// models/StudentFeedback.js
const mongoose = require('mongoose');

const feedbackSchema = new mongoose.Schema(
  {
    teacherId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    studentId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    subject: { type: String, required: true },
    ratings: {
      communication: { type: Number, min: 1, max: 5, required: true },
      clarity: { type: Number, min: 1, max: 5, required: true },
      engagement: { type: Number, min: 1, max: 5, required: true },
      knowledge: { type: Number, min: 1, max: 5, required: true },
      punctuality: { type: Number, min: 1, max: 5, required: true },
    },
    overallRating: { type: Number },
    comments: { type: String, maxlength: 1000 },
    semester: { type: String },
    academicYear: { type: String },
    sentiment: {
      label: { type: String, enum: ['positive', 'negative', 'neutral'] },
      score: { type: Number },
      keywords: [{ type: String }],
    },
  },
  { timestamps: true }
);

feedbackSchema.pre('save', function (next) {
  const r = this.ratings;
  this.overallRating =
    (r.communication + r.clarity + r.engagement + r.knowledge + r.punctuality) / 5;
  next();
});

// One feedback per student per teacher per subject per semester
feedbackSchema.index(
  { teacherId: 1, studentId: 1, subject: 1, semester: 1, academicYear: 1 },
  { unique: true }
);

module.exports = mongoose.model('StudentFeedback', feedbackSchema);
